### PR TITLE
CB-8: Add sing in/sing out ( clean one)

### PR DESCRIPTION
**Pull request by mirecki**
https://team-1626248039387.atlassian.net/jira/software/projects/CB/boards/1?selectedIssue=CB-8&text=BE

### CODE REVIEW

**Review by Narzew — APPROVED** [review]

**Review by grabomarcin — APPROVED** [review]

File: Gemfile [hunk 1]
```diff
@@ -27,11 +27,12 @@ gem 'bootsnap', '>= 1.4.4', require: false
 gem 'rack-cors', '~> 1.1.1'
 
 gem 'active_model_serializers', '~> 0.10'
+gem 'devise_token_auth', '~> 1.2.0'
 gem 'devise', '~> 4.8.0'
 gem 'rswag-api', '~> 2.4.0'
 gem 'rswag-ui', '~> 2.4.0'
 gem 'validate_url', '~> 1.0.13'
-
+gem 'omniauth', '~> 2.0', '>= 2.0.4'
```
Comment (h4b00) [inline Gemfile:35]: where do we use it?

File: app/models/user.rb [hunk 2]
```diff
@@ -3,39 +3,66 @@
...
+  has_many :comments, dependent: :destroy
+  has_many :recipes, dependent: :destroy
 
   # == Validations ==========================================================
 
-  validates :nickname, :email, uniqueness: true, presence: true
+  validates :email, presence: true, uniqueness: { case_sensitive: false }
+  # validates :nickname, presence: true, uniqueness: true
+  validates :nickname, presence: true, uniqueness: true, format: { with: /^[a-zA-Z0-9_.]*$/, multiline: true }
+  validate :validate_nickname
```
Comment (h4b00) [inline app/models/user.rb:65]: `(uniqueness: true)` is not enough?

File: app/models/user.rb [hunk 3]
```diff
@@ -3,39 +3,66 @@
...
 #  index_users_on_nickname              (nickname) UNIQUE
 #  index_users_on_reset_password_token  (reset_password_token) UNIQUE
+#  index_users_on_uid_and_provider      (uid,provider) UNIQUE
 #
-
 class User < ApplicationRecord
+  include DeviseTokenAuth::Concerns::User
+  extend Devise::Models
+  # temp and for being safe
+  before_save -> { skip_confirmation! }
```
Comment (h4b00) [inline app/models/user.rb:42]: ?

File: app/models/user.rb [hunk 4]
```diff
@@ -3,39 +3,66 @@
...
   # == Relationships ========================================================
+  has_many :comments, dependent: :destroy
+  has_many :recipes, dependent: :destroy
 
   # == Validations ==========================================================
 
-  validates :nickname, :email, uniqueness: true, presence: true
+  validates :email, presence: true, uniqueness: { case_sensitive: false }
+  # validates :nickname, presence: true, uniqueness: true
+  validates :nickname, presence: true, uniqueness: true, format: { with: /^[a-zA-Z0-9_.]*$/, multiline: true }
```
Comment (h4b00) [inline app/models/user.rb:64]: multiline?

File: app/models/user.rb [hunk 5]
```diff
@@ -46,4 +73,12 @@ class User < ApplicationRecord
   # == Class Methods ========================================================
 
   # == Instance Methods =====================================================
+  # check if nickname != email
+  def validate_nickname
+    errors.add(:nickname, :invalid) if User.where(email: nickname).exists?
+  end
+
+  def login
+    @login || nickname || email
```
Comment (h4b00) [inline app/models/user.rb:82]: you have `presence: true` on the nickname

File: db/migrate/20210806115824_devise_create_users.rb [hunk 6]
```diff
@@ -1,46 +0,0 @@
-# frozen_string_literal: true
```
Comment (h4b00) [inline db/migrate/20210806115824_devise_create_users.rb:1]: you can't delete any migration. Create a new one that would drop this table.

File: spec/factories/users.rb [hunk 7]
```diff
@@ -3,25 +3,48 @@
...
 #  index_users_on_reset_password_token  (reset_password_token) UNIQUE
+#  index_users_on_uid_and_provider      (uid,provider) UNIQUE
 #
 FactoryBot.define do
   factory :user do
-    sequence(:email) { |n| "email#{n}@gmail.com" }
-    sequence(:nickname) { |n| "nickname#{n}" }
-    password { Faker::Internet.password }
+    name { Faker::Name.name }
+    nickname { Faker::Internet.username(specifier: 3..30) }
```
Comment (h4b00) [inline spec/factories/users.rb:41]: username is unique, so you can't use faker for that.

File: spec/factories/users.rb [hunk 8]
```diff
@@ -3,25 +3,48 @@
...
+    email { Faker::Internet.unique.safe_email }
+    password { 'password123#' }
+    password_confirmation { 'password123#' }
+  end
+  factory :userCredentials do
+    login { Faker::randomElement([Faker::Internet.username(specifier: 3..30).unique, Faker::Internet.unique.safe_email]) }
+    password { Faker::Internet.password(min_length: 6, max_length: 128, mix_case: true, special_characters: true) }
```
Comment (h4b00) [inline spec/factories/users.rb:48]: Where do you use it?

**Review by h4b00 — CHANGES_REQUESTED** [review]
Missing spec validation for `nickname`, missing specs for relations inside the user model.
Use shoulda-matcher for these.

File: Gemfile (hunk 1 above)
Comment (mirecki) [reply Gemfile]: I wanted to use it but I am dum dum so it is just unnecessary , fixed now

File: app/models/user.rb (hunk 4 above)
Comment (mirecki) [reply app/models/user.rb]: fixed to be one line, no code validation from tutorial on my part

File: app/models/user.rb (hunk 5 above)
Comment (mirecki) [reply app/models/user.rb:80]: As it suppose to if I understand it correctly, agreed with FE to nickname and email be required during sing up. It is just that you can log in via nickname or email as user wishes to. So there is no situation where nickname is not present

File: spec/factories/users.rb (hunk 7 above)
Comment (mirecki) [reply spec/factories/users.rb]: added to be unique

File: spec/factories/users.rb (hunk 8 above)
Comment (mirecki) [reply spec/factories/users.rb]: fixed now since it is unnecessary code

File: app/models/user.rb (hunk 3 above)
Comment (mirecki) [reply app/models/user.rb]: I have went for lazy way of keeping confirmable stuff setup for future but not interfering with currently implemented and working. 
Fixed it properly to declare in environments localhost for mailer and implementation of set confirmation url to success to one , thing is that for whatever reason I cant make work "protect from forgery with null session" in app controller.
Was planning to do it and read why it is happening like that in separate CR when I wont be blocking other ppl with mine stuff.

File: spec/factories/users.rb (hunk 7 above)
Comment (mirecki) [reply spec/factories/users.rb]: fixed with sequence

File: app/models/user.rb (hunk 2 above)
Comment (mirecki) [reply app/models/user.rb:64]: ok left only uniqueness then

File: db/migrate/20210806115824_devise_create_users.rb (hunk 6 above)
Comment (mirecki) [reply db/migrate/20210806115824_devise_create_users.rb]: fixed

File: app/models/user.rb [hunk 9]
```diff
@@ -3,39 +3,62 @@
...
+#  index_users_on_confirmation_token    (confirmation_token) UNIQUE
 #  index_users_on_email                 (email) UNIQUE
 #  index_users_on_nickname              (nickname) UNIQUE
 #  index_users_on_reset_password_token  (reset_password_token) UNIQUE
+#  index_users_on_uid_and_provider      (uid,provider) UNIQUE
 #
-
 class User < ApplicationRecord
+  include DeviseTokenAuth::Concerns::User
+  extend Devise::Models
```
Comment (sebastiansito) [inline app/models/user.rb]: Let's move these to extensions

File: app/models/user.rb (hunk 2 above)
Comment (h4b00) [reply app/models/user.rb:61]: So why it's still here?

File: app/models/user.rb (hunk 5 above)
Comment (h4b00) [reply app/models/user.rb:78]: exactly so login in your case is always nickname.

File: spec/models/user_spec.rb [hunk 10]
```diff
@@ -3,33 +3,52 @@
...
 
 RSpec.describe User, type: :model do
+  describe 'relationships' do
+    it { is_expected.to have_many(:rates).dependent(true) }
+  end
+
   describe 'validations' do
     subject { FactoryBot.create(:user) }
-
+
```
Comment (h4b00) [inline spec/models/user_spec.rb:46]: ?

**Review by h4b00 — CHANGES_REQUESTED** [review]

File: db/migrate/20210818112907_devise_token_auth_create_users.rb [hunk 11]
```diff
@@ -0,0 +1,56 @@
+class DeviseTokenAuthCreateUsers < ActiveRecord::Migration[6.1]
```
Comment (sebastiansito) [inline db/migrate/20210818112907_devise_token_auth_create_users.rb:1]: We can have only one migration.

File: db/migrate/20210818112907_devise_token_auth_create_users.rb [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+  def change
```
Comment (sebastiansito) [inline db/migrate/20210818112907_devise_token_auth_create_users.rb:2]: We should try to update existing table instead

**Review by sebastiansito — APPROVED** [review]
Pre-approved

File: app/models/user.rb (hunk 5 above)
Comment (sebastiansito) [reply app/models/user.rb:79]: Unless you write `user.login` at some point somewhere. We could use it at some point to override.

**Review by sebastiansito — APPROVED** [review]

File: app/models/user.rb (hunk 2 above)
Comment (sebastiansito) [reply app/models/user.rb]: `uniqueness` won't check if someone has the same email as other user's nickname

**Review by h4b00 — APPROVED** [review]

**Pull request merged**